### PR TITLE
Improve 'source apiserver update --resource'

### LINKED ISSUE #565: source apiserver update --resource should follow the common update semantics for list like structures
I.e. the same as for `kn service update --env` or `kn trigger update --filter`. I.e. add new resources, overwrite existing, and remove when suffixed with `-`.

### PR DESCRIPTION
**Pull request by daisy-ycguo**
Fixes #565

## Proposed Changes

* Improve `source apiserver update --resource` to follow the common update semantics
* Support remove resource when suffixed with `-`

### CODE REVIEW

File: pkg/kn/commands/source/apiserver/flags.go [hunk 1]
```diff
@@ -41,27 +41,44 @@ type APIServerSourceUpdateFlags struct {
...
+func (f *APIServerSourceUpdateFlags) getUpdateAPIServerResourceArray() ([]resourceSpec, []resourceSpec, error) {
+	var added []resourceSpec
+	var removed []resourceSpec
+	for _, r := range f.Resources {
+		if strings.HasSuffix(r, "-") {
+			resourceSpec, err := getValidResource(r[:len(r)-1])
+			if err != nil {
+				return nil, nil, err
+			}
+			removed = append(removed, *resourceSpec)
```
Comment (navidshaikh) [inline pkg/kn/commands/source/apiserver/flags.go]: Lets get https://github.com/knative/client/pull/592 in first and update this PR accordingly.

Comment (navidshaikh) [reply pkg/kn/commands/source/apiserver/flags.go]: @daisy-ycguo : #592 is merged, please update the PR to use the shared utility to parse minus suffix.

Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/flags.go]: Thank you @navidshaikh 
resources at `apiserver update` are not key=value format. So it's not convenient to use ParseMinusSuffix here. But I added a common utility in parsing_helper.go called `AddListAndRemovalListFromArray` to parse minus suffix in an array.

File: pkg/kn/commands/source/apiserver/flags.go [hunk 2]
```diff
@@ -41,27 +42,45 @@ type APIServerSourceUpdateFlags struct {
...
+			return nil, nil, err
 		}
-		resourceList = append(resourceList, resourceRef)
+		removed = append(removed, *resourceSpec)
 	}
-	return &resourceList, nil
+	return added, removed, nil
 }
 
 func getValidResource(resource string) (*resourceSpec, error) {
```
Comment (maximilien) [inline pkg/kn/commands/source/apiserver/flags.go]: you should add comment for all code. I would rather see for public methods and functions but since you added above then should be consistent...

File: pkg/kn/commands/source/apiserver/flags.go [hunk 3]
```diff
@@ -95,9 +114,9 @@ func (f *APIServerSourceUpdateFlags) Add(cmd *cobra.Command) {
 		`The mode the receive adapter controller runs under:,
 "Ref" sends only the reference to the resource,
 "Resource" send the full resource.`)
-	cmd.Flags().StringSliceVar(&f.Resources,
+	cmd.Flags().StringArrayVar(&f.Resources,
 		"resource",
```
Comment (maximilien) [inline pkg/kn/commands/source/apiserver/flags.go]: Why does the name `resource` not match the variable `f.Resources`? Should this not be `"resources"`?

File: pkg/kn/commands/source/apiserver/flags_test.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+)
+
+func TestGetAPIServerResourceArray(t *testing.T) {
+	t.Run("get single apiserver resource", func(t *testing.T) {
+		createFlag := APIServerSourceUpdateFlags{
+			ServiceAccountName: "test-sa",
+			Mode:               "Ref",
+			Resources:          []string{"Service:serving.knative.dev/v1alpha1:true"},
+		}
+		created, _ := createFlag.getAPIServerResourceArray()
```
Comment (maximilien) [inline pkg/kn/commands/source/apiserver/flags_test.go]: why are you ignoring errors? Just assert that they are nil

File: pkg/kn/commands/source/apiserver/flags_test.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+package apiserver
+
+import (
+	"testing"
+
+	"gotest.tools/assert"
+)
+
+func TestGetAPIServerResourceArray(t *testing.T) {
+	t.Run("get single apiserver resource", func(t *testing.T) {
```
Comment (maximilien) [inline pkg/kn/commands/source/apiserver/flags_test.go]: I think since there are many combination you are testing you should offer a small comment summary of the table and what combination you are testing. This would even help you if you need to add or remove tests. Or seperate them into their own `t.Run` calls

File: pkg/util/parsing_helper_test.go [hunk 6]
```diff
@@ -92,3 +92,9 @@ func TestParseMinusSuffix(t *testing.T) {
 	assert.DeepEqual(t, expectedMap, inputMap)
 	assert.DeepEqual(t, expectedStringToRemove, stringToRemove)
 }
+
+func TestAddListAndRemovalListFromArray(t *testing.T) {
```
Comment (maximilien) [inline pkg/util/parsing_helper_test.go]: I would add tests for corner cases. For instance, 
1. add one item and remove none, 
2. add no items, remove one, 
3. add one item and remove one, 
4. add none and remove none.

File: docs/cmd/kn_source_apiserver_create.md [hunk 7]
```diff
@@ -26,7 +26,7 @@ kn source apiserver create NAME --resource RESOURCE --service-account ACCOUNTNAM
                                  "Ref" sends only the reference to the resource,
                                  "Resource" send the full resource. (default "Ref")
   -n, --namespace string         Specify the namespace to operate in.
-      --resource strings         Specification for which events to listen, in the format Kind:APIVersion:isController, e.g. Deployment:apps/v1:true.
+      --resource stringArray     Specification for which events to listen, in the format Kind:APIVersion:isController, e.g. Deployment:apps/v1:true.
```
[the remark this answers is not in the document]
Comment (daisy-ycguo) [reply docs/cmd/kn_source_apiserver_create.md]: `Deployment:apps/v1:true` is just the format of `Kind:APIVersion:isController`. APIVersion usually looks like:
- `serving.knative.dev/v1` or
- simply `v1`, only means kubernetes API V1.

I could change it to `v1` if it looks confused.

File: docs/cmd/kn_source_apiserver_update.md [hunk 8]
```diff
@@ -26,7 +26,7 @@ kn source apiserver update NAME --resource RESOURCE --service-account ACCOUNTNAM
                                  "Ref" sends only the reference to the resource,
                                  "Resource" send the full resource. (default "Ref")
   -n, --namespace string         Specify the namespace to operate in.
-      --resource strings         Specification for which events to listen, in the format Kind:APIVersion:isController, e.g. Deployment:apps/v1:true.
+      --resource stringArray     Specification for which events to listen, in the format Kind:APIVersion:isController, e.g. Deployment:apps/v1:true.
```
[the remark this answers is not in the document]
Comment (daisy-ycguo) [reply docs/cmd/kn_source_apiserver_update.md]: see my previous comments

File: pkg/eventing/sources/v1alpha1/apiserver_client.go [hunk 9]
```diff
@@ -156,6 +159,43 @@ func (b *APIServerSourceBuilder) Resources(resources []v1alpha1.ApiServerResourc
...
+		resources = []v1alpha1.ApiServerResource{}
+		b.apiServerSource.Spec.Resources = resources
+	}
+	resourceRef := v1alpha1.ApiServerResource{
+		APIVersion: version,
+		Kind:       kind,
+		Controller: isController,
+	}
+	b.apiServerSource.Spec.Resources = append(resources, resourceRef)
+	return b
```
[the remark this answers is not in the document]
Comment (daisy-ycguo) [reply pkg/eventing/sources/v1alpha1/apiserver_client.go]: I follow the same style in other eventing clients, e.g. [cronjob_client](https://github.com/knative/client/blob/master/pkg/eventing/sources/v1alpha1/cronjob_client.go#L153) and [apiserver_client](https://github.com/knative/client/blob/master/pkg/eventing/sources/v1alpha1/apiserver_client.go#L174). If you prefer to change the return, I will suggest to create a separate issue to track and we can change them all together.

File: pkg/eventing/sources/v1alpha1/apiserver_client.go [hunk 10]
```diff
@@ -156,6 +159,43 @@ func (b *APIServerSourceBuilder) Resources(resources []v1alpha1.ApiServerResourc
...
+		Kind:       kind,
+		Controller: isController,
+	}
+	for i, k := range resources {
+		if reflect.DeepEqual(k, resourceRef) {
+			resources = append(resources[:i], resources[i+1:]...)
+			return b, nil
+		}
+	}
+	return b, fmt.Errorf("cannot find resource %s:%s:%t to remove", version, kind, isController)
```
[the remark this answers is not in the document]
Comment (daisy-ycguo) [reply pkg/eventing/sources/v1alpha1/apiserver_client.go]: If the resource could be found, the function will return at the line #193 without error.
If the resource could not be found after traversing the whole structure, the function will return at line #196 with an error.
Could you help to explain more of your comment ?

File: pkg/eventing/sources/v1alpha1/apiserver_client.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (daisy-ycguo) [reply pkg/eventing/sources/v1alpha1/apiserver_client.go]: b is short for builder. Again, I will suggest to create a separate issue to change all client codes, including  cronjob_client, eventing_client and apiserver_client.

File: pkg/kn/commands/source/apiserver/flags.go (hunk 3 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/flags.go]: `resource` is a flag. `Resources` is an array of `resource`.
User uses `--resource abc --resource def` to set multiple resources, which will be saved in `f.Resources`.

File: pkg/kn/commands/source/apiserver/flags_test.go (hunk 4 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/flags_test.go]: will add some tests to test errors. Thank you for pointing out.

File: pkg/kn/commands/source/apiserver/flags_test.go (hunk 5 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/flags_test.go]: simple comments are added before test cases. e.g. line 38 and line 52.
But I think maybe separate them into a single t.Run is more readable.

File: pkg/util/parsing_helper_test.go (hunk 6 above)
Comment (daisy-ycguo) [reply pkg/util/parsing_helper_test.go]: will do.

File: pkg/kn/commands/source/apiserver/flags.go [hunk 11]
```diff
@@ -41,27 +42,45 @@ type APIServerSourceUpdateFlags struct {
...
-		resourceRef := v1alpha1.ApiServerResource{
-			APIVersion: resourceSpec.apiVersion,
-			Kind:       resourceSpec.kind,
-			Controller: resourceSpec.isController,
+		resourceList = append(resourceList, *resourceSpec)
+	}
+	return resourceList, nil
+}
+
+// getAPIServerResourceArray is to return an array of ApiServerResource from a string. A sample is Event:v1:true
```
[the remark this answers is not in the document]
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/flags.go]: will do

File: pkg/eventing/sources/v1alpha1/apiserver_client.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (rhuss) [reply pkg/eventing/sources/v1alpha1/apiserver_client.go]: This is classical pattern used in many languages which allows for chained calls for elegant, DSL-like APIs. Do you remember the review comment where there was a check for a nil-received in a method (in some dependency) ? That is exactly for that use case of chained calls, where an intermediate call could return nil. So its also used in Go quite a bit (and btw is very popular in Java as well).

Some references illustrating that chaining is quite popular in Go, too (and they all return the object itself):

* https://github.com/mauricioklein/go-chainable
* https://vikash1976.wordpress.com/2017/02/03/method-chaining-in-go-lang-through-interface/
* https://riptutorial.com/go/example/13465/chaining-methods
* https://stackoverflow.com/questions/27297702/go-method-chaining-and-error-handling
* ....

BTW, you can still ignore the returned value if you like.

File: pkg/kn/commands/source/apiserver/flags.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/flags.go]: done.

File: pkg/kn/commands/source/apiserver/flags.go (hunk 2 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/flags.go]: Added a comment.

**Comment by maximilien** [comment]
Is this ready? Also, looks like you have some conflicts to resolve. Thanks.

File: pkg/util/parsing_helper_test.go (hunk 6 above)
Comment (daisy-ycguo) [reply pkg/util/parsing_helper_test.go]: done.

File: docs/cmd/kn_source_apiserver_create.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (rhuss) [reply docs/cmd/kn_source_apiserver_create.md]: I would keep `apps/v1` as `v1` without an apigroup is more the exception than the rule. Giving a full example makes more sense.

File: pkg/kn/commands/source/apiserver/flags_test.go (hunk 5 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/flags_test.go]: done.

File: pkg/eventing/sources/v1alpha1/apiserver_client.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (rhuss) [reply pkg/eventing/sources/v1alpha1/apiserver_client.go]: @maximilien  is right, chaining on works on methods which return a single return value, so returning two values makes no sense in this pattern.

But as we discussed in the other PR having mutating methods on a builder is an anti-pattern (as then the builder turns more into a regular domain objects). Maybe we can solve this use case differently, too ?

If not, we could agree that removing a non-existing resource is just a no-op and gets ignored, so that you can return directly the builder again.

File: pkg/kn/commands/source/apiserver/update.go [hunk 12]
```diff
@@ -74,11 +74,19 @@ func NewAPIServerUpdateCommand(p *commands.KnParams) *cobra.Command {
...
-				b.Resources(*resources)
+				for _, k := range removed {
+					_, err = b.RemoveResource(k.ApiVersion, k.Kind, k.IsController)
+					if err != nil {
+						return err
+					}
+				}
+				for _, k := range added {
+					b.AddResource(k.ApiVersion, k.Kind, k.IsController)
+				}
```
Comment (rhuss) [inline pkg/kn/commands/source/apiserver/update.go]: I would follow the same scheme here as we had for trigger filters in #603 , I think it covers the same use case.

Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/update.go]: ok. will do.

Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/update.go]: Resource is not a string map, but an array. I cannot leverage stringmap here.
I have to think about how to handle.

File: pkg/kn/commands/source/apiserver/flags.go [hunk 13]
```diff
@@ -40,47 +42,99 @@ type APIServerSourceUpdateFlags struct {
...
-			APIVersion: resourceSpec.apiVersion,
-			Kind:       resourceSpec.kind,
-			Controller: resourceSpec.isController,
+		resourceList = append(resourceList, *resourceSpec)
+	}
+	return resourceList, nil
+}
+
+// updateExistingAPIServerResourceArray is to update an array of resources.
+func (f *APIServerSourceUpdateFlags) updateExistingAPIServerResourceArray(existing []v1alpha1.ApiServerResource) ([]v1alpha1.ApiServerResource, error) {
```
Comment (daisy-ycguo) [inline pkg/kn/commands/source/apiserver/flags.go]: I would like to move this method to parsing_helper.go to provide a common method to update existing array.
My question is which type (or key word) I should use to describe a general array here ?

I tried to use []interface{} but I get an error:
```
cannot use existing (type []"knative.dev/eventing/pkg/apis/sources/v1alpha1".ApiServerResource) as type []interface {} in argument to util.UpdateArray
```

Comment (rhuss) [reply pkg/kn/commands/source/apiserver/flags.go]: That's not easily possible, see https://golang.org/doc/faq#convert_slice_of_interface

File: pkg/kn/commands/source/apiserver/update.go (hunk 12 above)
Comment (rhuss) [reply pkg/kn/commands/source/apiserver/update.go]: Ah, right. Unfortunately, there are no generics in golang, that would make things much easier. Its also not easy to simulate with interfaces ;(

File: pkg/kn/commands/source/apiserver/flags.go (hunk 13 above)
Comment (rhuss) [reply pkg/kn/commands/source/apiserver/flags.go]: What you could do is to introduce a simple utility method which converts an `[]ApiServerResource` to a `[]interface{}` and feed that into the generic method like in 

```golang
fromISlice(UpdateArray(toISlice(existing), toISliceP(added), toISliceP(removed)))
```

and do the conversion between ApiServerSource and interfaces slices before and after calling the function. But I don't think its worth the hassle.

Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/flags.go]: oh, ok. Then I keep as it is. 
Please review the latest version. 
Thank you.

**Comment by daisy-ycguo** [comment]
> Is this ready? Also, looks like you have some conflicts to resolve. Thanks.

Yes, it's ready for a review.

File: pkg/kn/commands/source/apiserver/flags.go [hunk 14]
```diff
@@ -40,47 +42,99 @@ type APIServerSourceUpdateFlags struct {
...
+
+	addedArray, removedArray := util.AddedAndRemovalListsFromArray(f.Resources)
+	for _, r := range addedArray {
+		resourceSpec, err := getValidResource(r)
+		if err != nil {
+			return nil, nil, err
+		}
+		added = append(added, *resourceSpec)
+	}
+	for _, r := range removedArray {
```
Comment (rhuss) [inline pkg/kn/commands/source/apiserver/flags.go]: you could put the loop into a seperate function and call it twice, for adds and removals. That would reduce duplicated code a bit.

File: pkg/kn/commands/source/apiserver/flags.go [hunk 15]
```diff
@@ -40,47 +42,99 @@ type APIServerSourceUpdateFlags struct {
...
+//getValidResource is to parse resource spec from a string
+func getValidResource(resource string) (*v1alpha1.ApiServerResource, error) {
+	var isController bool
+	var version string
 	var err error
 
-	parts := strings.Split(resource, apiVersionSplitChar)
+	parts := strings.SplitN(resource, apiVersionSplitChar, 3)
+	if len(parts[0]) == 0 {
+		return nil, fmt.Errorf("cannot find KIND in resource specification %s", resource)
```
Comment (rhuss) [inline pkg/kn/commands/source/apiserver/flags.go]:
```suggestion
		return nil, fmt.Errorf("cannot find 'Kind' part in resource specification %s (expected: <Kind:ApiVersion[:controllerFlag]>", resource)
```

File: pkg/kn/commands/source/apiserver/flags.go [hunk 16]
```diff
@@ -40,47 +42,99 @@ type APIServerSourceUpdateFlags struct {
...
+	}
 	kind := parts[0]
-	if len(parts) < 2 {
-		return nil, fmt.Errorf("no APIVersion given for resource %s", resource)
+
+	if len(parts) >= 2 && len(parts[1]) > 0 {
+		version = parts[1]
+	} else {
+		version = "v1"
```
Comment (rhuss) [inline pkg/kn/commands/source/apiserver/flags.go]: Do we really want a default to "v1" ? I would prefer to be always explicit with the ApiVersion and make it mandatory (as you have to do it for any k8s resource, too, when deploying to the server)

File: pkg/kn/commands/source/apiserver/flags.go [hunk 17]
```diff
@@ -40,47 +42,99 @@ type APIServerSourceUpdateFlags struct {
...
 	}
-	version := parts[1]
-	if len(parts) >= 3 {
+	if len(parts) >= 3 && len(parts[2]) > 0 {
 		isController, err = strconv.ParseBool(parts[2])
 		if err != nil {
 			return nil, fmt.Errorf("cannot parse controller flage in resource specification %s", resource)
```
Comment (rhuss) [inline pkg/kn/commands/source/apiserver/flags.go]:
```suggestion
			return nil, fmt.Errorf("controller flag is not a boolean in resource specification %s (expected: <Kind:ApiVersion[:controllerFlag]>)", resource)
```

File: pkg/kn/commands/source/apiserver/flags.go (hunk 14 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/flags.go]: done

File: pkg/kn/commands/source/apiserver/flags.go (hunk 15 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/flags.go]: done

File: pkg/kn/commands/source/apiserver/flags.go (hunk 17 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/flags.go]: done

File: pkg/kn/commands/source/apiserver/flags.go (hunk 16 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/flags.go]: I just want to make a simple format for ApiServerSource, e.g. Event or Pod or Deployment could only be like `--resource Event --resource Pod --resource Deployment`, other than `--resource Event:v1 --resource Pod:v1 --resource Deployment:v1` which doesn't look artistic.

So if no ApiVersion provides, I use v1 instead. If you don't agree with this kind of contraction rule, I could delete this function.

Comment (rhuss) [reply pkg/kn/commands/source/apiserver/flags.go]: My point is that this is a bit surprising for Kubernetes users, as the apiVersion is a mandatory field in a resource descriptors. Also, imagine now that the Event or Pod spec evolves to a v2 version. Would you then also update the default to get to this newest version ? Will a user still using v1 and, more important, is she aware of that not the latest event version is used ? Maybe even v1 is getting removed later on then, making the default useless.

I understand that you want to make it as easy and concise as possible. However, its one of  key principle of Kubernetes (and hence Knative, too) that to identify a resource you alway need apiVersion + kind. 

So I really would be explicit here to (a) be on par with the Kubernetes spec and (b) to avoid future surprises (and (c) to avoid 'hacks' like '::' to get to the third option).

Comment (rhuss) [reply pkg/kn/commands/source/apiserver/flags.go]: btw, `Deployment` is in `apps/v1`, not `v1`. There are only a handful of resources that are really in v1.

Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/flags.go]: oh. Got your point. Let me remove the default value.

Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/flags.go]: Done. The latest version is submitted and ready for review. @rhuss

**Comment by daisy-ycguo** [comment]
@maximilien @rhuss Are we ready to merge this ? I'm on vacation. I could still address any comments if you have. Thank you.
